11_FFTBloom/main.cpp

```diff
@@ -0,0 +1,1192 @@
+
+
+// TODO: Copy HelloSwapchain to set up a swapchain, draw to a surface. Or copy the new MPMC example.
+// TODO: Dynamically modify bloom size with mouse scroll/ +- keys
+//		 Notes: To do so must check required size against buffer size (expand buffer and recompile kernel if it starts being small), also after making kernel larger then small again
+//		        there's probably no need to shrink buffer and recompile kernel but would be nice to add that as well)
+// TODO: Refactor after FFT ext is back
+// TODO: Make sampling formats be #defined depending on how they were loaded on GPU side
+
+using namespace nbl;
+using namespace core;
+using namespace system;
+using namespace asset;
+using namespace video;
+
+
+#include "app_resources/common.hlsl"
+#include "nbl/builtin/hlsl/bit.hlsl"
+#include "utils.h"
+
+// In this application we'll cover buffer streaming, Buffer Device Address (BDA) and push constants 
+class FFTBloomApp final : public application_templates::MonoDeviceApplication, public application_templates::MonoAssetManagerAndBuiltinResourceApplication
+{
+	using device_base_t = application_templates::MonoDeviceApplication;
+	using asset_base_t = application_templates::MonoAssetManagerAndBuiltinResourceApplication;
+
+	// Compute Pipelines
+	smart_refctd_ptr<IGPUComputePipeline> m_firstAxisFFTPipeline;
+	smart_refctd_ptr<IGPUComputePipeline> m_lastAxisFFT_convolution_lastAxisIFFTPipeline;
+	smart_refctd_ptr<IGPUComputePipeline> m_firstAxisIFFTPipeline;
+
+	// Descriptor Sets
+	smart_refctd_ptr<IGPUDescriptorSet> m_firstAxisFFTDescriptorSet;
+	smart_refctd_ptr<IGPUDescriptorSet> m_lastAxisFFT_convolution_lastAxisIFFTDescriptorSet;
+	smart_refctd_ptr<IGPUDescriptorSet> m_firstAxisIFFTDescriptorSet;
+
+	// Utils
+	smart_refctd_ptr<IUtilities> m_utils;
+
+	// Resources
+	smart_refctd_ptr<IGPUImageView> m_srcImageView;
+	smart_refctd_ptr<IGPUImageView> m_kerImageView;
+	smart_refctd_ptr<IGPUImage> m_outImg;
+	smart_refctd_ptr<IGPUImageView> m_outImgView;
+	smart_refctd_ptr<IGPUImageView> m_kernelNormalizedSpectrums[CHANNELS];
+
+	// Used to store intermediate results
+	smart_refctd_ptr<IGPUBuffer> m_rowMajorBuffer;
+	smart_refctd_ptr<IGPUBuffer> m_colMajorBuffer;
+
+	// These are Buffer Device Addresses
+	uint64_t m_rowMajorBufferAddress;
+	uint64_t m_colMajorBufferAddress;
+
+	// Some parameters
+	float bloomScale = 1.f;
+	float useHalfFloats = false;
+	
+	// Other parameter-dependent variables
+	asset::VkExtent3D marginSrcDim;
+
+	// We only hold onto one cmdbuffer at a time, but having it here doesn't hurt
+	smart_refctd_ptr<IGPUCommandBuffer> m_computeCmdBuf;
+
+	// Shader Cache
+	smart_refctd_ptr<IShaderCompiler::CCache> m_readCache;
+	smart_refctd_ptr<IShaderCompiler::CCache> m_writeCache;
+
+	// Sync primitives
+	smart_refctd_ptr<ISemaphore> m_timeline;
+	uint64_t semaphorValue = 0;
+	// For image uploads
+	SIntendedSubmitInfo m_intendedSubmit;
+	smart_refctd_ptr<ISemaphore> m_scratchSemaphore;
```

Comment:
you don't need to keep it around as a member, the queue will hold onto it past the dispatch and it will be alive until it signals


A lot of stuff you don't actually need to hold onto as a member because of refcounting and lifetime tracking